### PR TITLE
support mysql binlog sync to pulsar by canal

### PR DESCRIPTION
**Pull request by tuteng**
### Motivation

support alibaba canal
https://github.com/alibaba/canal/wiki

### Modifications

Integrated canal client

### Result

support binlog sync to pulsar
use python pulsar-client to consume

```
import pulsar

client = pulsar.Client('pulsar://localhost:6650')
consumer = client.subscribe('my-topic',
                            subscription_name='my-sub')

while True:
    msg = consumer.receive()
    print("Received message: '%s'" % msg.data())
    consumer.acknowledge(msg)

client.close()
```

```
output:
Received message: '[{"data":null,"database":"testdb","es":1542446501000,"id":44,"isDdl":true,"mysqlType":null,"old":null,"sql":"CREATE TABLE `users320` (   `id` int(11) NOT NULL AUTO_INCREMENT,   `name` varchar(50) DEFAULT NULL,   `extra` varchar(50) DEFAULT NULL,   PRIMARY KEY (`id`),   KEY `ix_users_name` (`name`) ) ENGINE=InnoDB DEFAULT CHARSET=utf8","sqlType":null,"table":"users320","ts":1542446501114,"type":"CREATE"}]'
```

### CODE REVIEW

File: pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                    canalSourceConfig.getDestination(), canalSourceConfig.getUsername(), canalSourceConfig.getPassword());
+        }
+        log.info("start canal connect");
+        this.start();
+
+    }
+
+    protected void start() {
+        Assert.notNull(connector, "connector is null");
+        thread = new Thread(new Runnable() {
```
Comment (merlimat) [inline pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java]: Maybe assign a name to thread so it's easier to debug with jstack

File: pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
+package org.apache.pulsar.io.canal;
```
Comment (merlimat) [inline pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java]: This needs to have Apache headers. You can use `mvn license:format` to automatically add

File: pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    private CanalConnector connector;
+
+    private CanalSourceConfig canalSourceConfig;
+
+    protected Thread.UncaughtExceptionHandler handler = new Thread.UncaughtExceptionHandler() {
+
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+            log.error("parse events has an error", e);
```
Comment (sijie) [inline pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java]: nit: it is good to add thread name in the logging message for troubleshooting purpose.

File: pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    protected Thread thread = null;
+
+    protected volatile boolean running = false;
+
+    private CanalConnector connector;
+
+    private CanalSourceConfig canalSourceConfig;
+
+    protected Thread.UncaughtExceptionHandler handler = new Thread.UncaughtExceptionHandler() {
```
Comment (sijie) [inline pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java]: `protected final`

File: pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        canalSourceConfig = CanalSourceConfig.load(config);
+        if (canalSourceConfig.getCluster()) {
+            connector = CanalConnectors.newClusterConnector(canalSourceConfig.getZkServers(),
+                    canalSourceConfig.getDestination(), canalSourceConfig.getUsername(), canalSourceConfig.getPassword());
+        } else {
+            connector = CanalConnectors.newSingleConnector(
+                    new InetSocketAddress(canalSourceConfig.getSingleHostname(), canalSourceConfig.getSinglePort()),
+                    canalSourceConfig.getDestination(), canalSourceConfig.getUsername(), canalSourceConfig.getPassword());
+        }
+        log.info("start canal connect");
```
Comment (sijie) [inline pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java]: can you improve the logging here?

- in cluster mode, can you print out the zookeeper user and destination?

- in single mode, can you print out the single host name, port and destination?

File: pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        this.start();
+
+    }
+
+    protected void start() {
+        Assert.notNull(connector, "connector is null");
```
Comment (sijie) [inline pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java]: nit: Use java builtin `objects.requirenonnull` or guava checkNotNull?

File: pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        running = false;
+        if (thread != null) {
+            thread.interrupt();
+            thread.join();
+        }
+        if (connector != null) {
+            connector.disconnect();
+        }
+
+        MDC.remove("destination");
```
Comment (sijie) [inline pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java]: can you define `destination` as a constant string?

File: pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+                MDC.remove("destination");
+            }
+        }
+    }
+
+
+    static private class CanalRecord implements Record<byte[]> {
+
+        private final byte[] record;
+        private final Long id;
```
Comment (sijie) [inline pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java]: is this id incremental? if you might consider using it as record sequence.

File: pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                    } else {
+                        if (flatMessages != null) {
+                            for (FlatMessage flatMessage : flatMessages) {
+                                String m = JSON.toJSONString(flatMessage, SerializerFeature.WriteMapNullValue);
+                                consume(new CanalRecord(m.getBytes(), batchId));
+                            }
+                        }
+                    }
+
+                    connector.ack(batchId);
```
Comment (sijie) [inline pulsar-io/canal/src/main/java/org/apache/pulsar/io/canal/CanalSource.java]: if you are acking here, it only supports at-most-once semantic. you might consider implementing ack() in CanalRecord

File: pulsar-io/pom.xml [hunk 10]
```diff
@@ -45,6 +45,7 @@
     <module>elastic-search</module>
     <module>kafka-connect-adaptor</module>
     <module>debezium</module>
+      <module>canal</module>
```
Comment (sijie) [inline pulsar-io/pom.xml]: nit: fix the indent here

**Comment by tuteng** [comment]
Please take the time to review my code again. It has been changed and tested locally.

**Comment by sijie** [comment]
run integration tests

**Comment by jiazhai** [comment]
run integration tests

**Comment by jiazhai** [comment]
run integration tests

**Comment by jiazhai** [comment]
run integration tests

**Comment by jiazhai** [comment]
run integration tests

**Comment by jiazhai** [comment]
run integration tests

**Comment by jiazhai** [comment]
run integration tests

**Comment by sijie** [comment]
@tuteng thank you so much for your contribution.

**Pull request merged**